### PR TITLE
Enhance X.509 CRT negative parsing tests

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR modifies the X.509 CRT negative parsing tests to always use certificates which are ill-formed in exactly one place, but well-formed apart from that. The purpose of this is to pave the way for upcoming changes in the parsing routine which will likely move from depth-first to (on-demand) breadth-first parsing. With breadth-first parsing, the current tests don't work anymore as the raw CRT data they use stops after the ill-formed entry under test, while breadth-first parsing requires the property that when a field within an ASN.1 substructure is ill-formed, the parent structure of that substructure should be well-formed. Furthermore, providing more data after the ill-formed field helps to catch bugs like #2437 regarding parsing functions not properly enforcing ASN.1 substructure bounds.

In addition to using full certificates when possible, this PR adds numerous test cases exercising the various ways an ASN.1 field can be ill-formed: Bad tag, invalid length encoding, length data missing, length out of bounds, length mismatches, ... It does not (yet?) attempt to be complete in that regard, exercising all these options for all CRT fields, but I made improvements where I saw them.

This PR is based on #2442.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Here's a little script which calls `dumpasn1` on each `x509parse_crt` test in `test_suite_x509parse.data` and inserts its output - maybe it's helpful for reviewing:
```bash
#!/bin/bash

HEX=$(echo $1 | sed -n 's/^x509parse_crt:\"\([a-zA-Z0-9]*\)\".*$/\1/p')
if [ -n "$HEX" ]; then
    echo "-------------------------------- ASN.1 DUMP ---------------------------------"
    echo $HEX | xxd -r -p - - > _tmp_out
    BYTES=$( wc -c _tmp_out | awk '{print $1}' )
    echo "DUMP: $HEX ($BYTES Bytes)"
    dumpasn1 -v _tmp_out 2>&1
    echo "-----------------------------------------------------------------------------"
fi
echo $1
```

**Comment by hanno-arm** [comment]
Tests exercising the X.509 CRT parser's behaviour on unexpected tags now use a valid ASN.1 NULL structure. That should ease reviewing because each block of tests concerning a particular field begins with an invalid-tag test that now uses valid ASN.1 data, hence can be displayed by the parser and easily inspected to test the respective field. The remaining tests are then usually small deviations from that test.

File: tests/suites/test_suite_x509parse.data [hunk 1]
```diff
@@ -942,47 +942,65 @@ x509parse_crt:"30153000300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X5
...
-x509parse_crt:"30163014a0030201048204deadbeef30070604cafed00d01":"":MBEDTLS_ERR_X509_INVALID_ALG + MBEDTLS_ERR_ASN1_OUT_OF_DATA
+x509parse_crt:"30163014a0030201028204deadbeef30070604cafed00d01":"":MBEDTLS_ERR_X509_INVALID_ALG + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
 X509 Certificate ASN1 (TBSCertificate, correct alg data, no params unknown version)
-x509parse_crt:"30153013a0030201048204deadbeef30060604cafed00d":"":MBEDTLS_ERR_X509_UNKNOWN_VERSION
+x509parse_crt:"30153013a0030201028204deadbeef30060604cafed00d":"":MBEDTLS_ERR_X509_UNKNOWN_VERSION
 
 X509 Certificate ASN1 (TBSCertificate, correct alg data, unknown version)
-x509parse_crt:"30173015a0030201048204deadbeef30080604cafed00d0500":"":MBEDTLS_ERR_X509_UNKNOWN_VERSION
+x509parse_crt:"30173015a0030201028204deadbeef30080604cafed00d0500":"":MBEDTLS_ERR_X509_UNKNOWN_VERSION
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Why change version from value 4 to value 2? Aren't both unknown?

File: tests/suites/test_suite_x509parse.data [hunk 2]
```diff
@@ -987,40 +987,45 @@ x509parse_crt:"301c3007a0030201028285300d06092a864886f70d01010b0500030200ff":"":
...
+X509 Certificate ASN1 (TBSCertificate, invalid AlgorithmIdentifier, OID empty)
+x509parse_crt:"30818b308175a0030201008204deadbeef30020600300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_UNKNOWN_SIG_ALG + MBEDTLS_ERR_OID_NOT_FOUND
 
-X509 Certificate ASN1 (TBSCertificate, correct alg, unknown alg_id)
-x509parse_crt:"30173015a0030201028204deadbeef30080604cafed00d0500":"":MBEDTLS_ERR_X509_UNKNOWN_SIG_ALG + MBEDTLS_ERR_OID_NOT_FOUND
+X509 Certificate ASN1 (TBSCertificate, invalid AlgorithmIdentifier, OID unknown)
+x509parse_crt:"30818f308179a0030201008204deadbeef30060604deadbeef300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_UNKNOWN_SIG_ALG + MBEDTLS_ERR_OID_NOT_FOUND
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: The OID of DEADBEEF is only 4 bytes long, which makes it an invalid not unknown OID. Consider using any 9 byte long OID, like DEADBEEF:DEADBEEF:42

File: tests/suites/test_suite_x509parse.data [hunk 3]
```diff
@@ -1122,21 +1122,89 @@ X509 Certificate ASN1 (TBSCertificate, invalid Issuer, 2nd AttributeTypeValue em
...
+
+X509 Certificate ASN1 (TBSCertificate, invalid Validity, notAfter invalid)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"308196308180a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303931323331323335393539170c303930313031303000000000300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_DATE
+
+X509 Certificate ASN1 (TBSCertificate, invalid Validity, notAfter data remaining)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"308198308182a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301e170c303930313031303030303030170c3039313233313233353935391700300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_DATE + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: "notAfter data remaining" I'm assuming to mean we have a too large of notAfter? If so, should we increase the length field beyond what a date would have and then add some more data for the notAfter value?

What this appears to be testing now is an extra zero-length date added after notAfter, which may or may not be what you want to test.

File: tests/suites/test_suite_x509parse.data [hunk 4]
```diff
@@ -1636,6 +1636,38 @@ X509 Certificate ASN1 (invalid SignatureAlgorithm, not the same as SignatureAlgo
...
+
+X509 Certificate ASN1 (invalid Signature, length out of bounds)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"3081bb3081a7a0030201028204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffffa100a200a321301f301d0603551d11041630148208666f6f2e7465737482086261722e74657374300d06092a864886f70d01010b05000301":"":MBEDTLS_ERR_X509_INVALID_SIGNATURE + MBEDTLS_ERR_ASN1_OUT_OF_DATA
+
+X509 Certificate ASN1 (invalid Signature, empty)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"3081bb3081a7a0030201028204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffffa100a200a321301f301d0603551d11041630148208666f6f2e7465737482086261722e74657374300d06092a864886f70d01010b05000300":"":MBEDTLS_ERR_X509_INVALID_SIGNATURE + MBEDTLS_ERR_ASN1_INVALID_DATA
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Do we want to employ a missing element for "empty" as written? Do also want to test a zero-length non-missing element for the signature?

File: tests/suites/test_suite_x509parse.data [hunk 5]
```diff
@@ -907,7 +907,7 @@ X509 CRT ASN1 (Empty Certificate)
 x509parse_crt:"":"":MBEDTLS_ERR_X509_INVALID_FORMAT
 
 X509 CRT ASN1 (inv Certificate, bad tag)
-x509parse_crt:"00":"":MBEDTLS_ERR_X509_INVALID_FORMAT
+x509parse_crt:"0500":"":MBEDTLS_ERR_X509_INVALID_FORMAT
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Aren't we reducing test coverage by only testing with ASN.1 structures that are well-formed? Shouldn't we still continue testing with ill-formed ASN.1 structures?

File: tests/suites/test_suite_x509parse.data [hunk 6]
```diff
@@ -1306,9 +1306,9 @@ X509 CRT ASN1 (TBS, SubPubKeyInfo missing)
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"30693054a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_PK_KEY_INVALID_FORMAT + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
-X509 CRT ASN1 (TBS, inv SubPubKeyInfo, inv tag
+X509 CRT ASN1 (TBS, inv SubPubKeyInfo, inv tag)
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: This should be moved to commit "Fix test case name formatting in X.509 parsing suite", shouldn't it? It doesn't have to do with "Use ASN.1 NULL TLVs when testing invalid tags ".

File: tests/suites/test_suite_x509parse.data [hunk 7]
```diff
@@ -1326,9 +1326,9 @@ X509 CRT ASN1 (TBS, inv SubPubKeyInfo, empty
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"306b3056a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a300806001304546573743000300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_PK_INVALID_ALG + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
-X509 CRT ASN1 (TBS, inv SubPubKeyInfo, inv algorithm tag
+X509 CRT ASN1 (TBS, inv SubPubKeyInfo, inv algorithm tag)
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Same here (move to commit "Fix test case name formatting in X.509 parsing suite")

File: tests/suites/test_suite_x509parse.data [hunk 8]
```diff
@@ -1354,9 +1354,9 @@ X509 CRT ASN1 (TBS, inv SubPubKeyInfo, bitstring missing
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"307a3065a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374300f300d06092A864886F70D0101010500300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_PK_INVALID_PUBKEY + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
-X509 CRT ASN1 (TBS, inv SubPubKeyInfo, bitstring inv tag
+X509 CRT ASN1 (TBS, inv SubPubKeyInfo, bitstring inv tag)
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Same here (move to commit "Fix test case name formatting in X.509 parsing suite")

File: tests/suites/test_suite_x509parse.data [hunk 9]
```diff
@@ -367,883 +367,1401 @@ X509 Time Future #6
...
+x509parse_crt:"3085":"":MBEDTLS_ERR_X509_INVALID_FORMAT
+
+X509 CRT ASN1 (inv Certificate, length data incomplete)
+x509parse_crt:"308200":"":MBEDTLS_ERR_X509_INVALID_FORMAT
+
+X509 CRT ASN1 (inv Certificate, length out of bounds)
+x509parse_crt:"3001":"":MBEDTLS_ERR_X509_INVALID_FORMAT
+
+X509 CRT ASN1 (inv TBS, invalid tag)
+x509parse_crt:"30020500":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: Is the "00" at the end needed here? I mean - you added it to have a valid NULL field, but shouldn't we fail regardless of whether anything follows it or not?

File: tests/suites/test_suite_x509parse.data [hunk 10]
```diff
@@ -367,883 +367,1401 @@ X509 Time Future #6
...
 
-X509 Certificate ASN1 (Correct first tag, length data incomplete)
-x509parse_crt:"3003308200":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_OUT_OF_DATA
+X509 CRT ASN1 (TBS empty)
+depends_on:MBEDTLS_SHA256_C:MBEDTLS_RSA_C
+x509parse_crt:"30153000300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
-X509 Certificate ASN1 (Correct first tag, second tag no TBSCertificate)
-x509parse_crt:"300100":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
+X509 CRT ASN1 (TBS, invalid version tag, serial missing)
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: Why does this test two problems at once?

File: tests/suites/test_suite_x509parse.data [hunk 11]
```diff
@@ -367,883 +367,1401 @@ X509 Time Future #6
...
+x509parse_crt:"30293014a012021100000000000000000000000000000000300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_VERSION + MBEDTLS_ERR_ASN1_OUT_OF_DATA
 
-X509 Certificate ASN1 (TBSCertificate, no version tag, serial missing)
-x509parse_crt:"3003300100":"":MBEDTLS_ERR_X509_INVALID_SERIAL + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG
+X509 CRT ASN1 (TBS, valid version tag + length, unknown version number 3)
+depends_on:MBEDTLS_SHA256_C:MBEDTLS_RSA_C
+x509parse_crt:"308196308180a0030201038204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_UNKNOWN_VERSION
 
-X509 Certificate ASN1 (TBSCertificate, invalid version tag)
-x
```
Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: How does this test differ from the previous one?

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: This test has an invalid issuer, and it shouldn't.

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: This test has the Sequence tag of RDNSequence of "00" instead of "30", which might be interpreted as the previous data too long instead of an invalid tag, since, as far as I know, 00, won't be recognized as any tag. Shouldn't we have, for example, 05 00 here instead?

File: tests/suites/test_suite_x509parse.data (hunk 10 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: It doesn't - the version is optional, so if no valid version tag is found, the parser will try to interpret the bytes as the CRT serial number. That's why we're expecting an `INVALID_SERIAL` error code here. What about renaming the test to: "TBS, invalid first tag, neither version nor serial"?

File: tests/suites/test_suite_x509parse.data (hunk 11 above)
Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: Length is not missing in this test, it is incorrect though.

File: tests/suites/test_suite_x509parse.data (hunk 9 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: I always used `0500` (which is a valid ASN.1 `NULL` TLV) for unexpected tag testing; the reason is that this way the ASN.1 data is still well-formed and can easily be displayed in a generic parser.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes we should, but I used `0500` for all invalid tag tests to have them use well-formed ASN.1 data.

File: tests/suites/test_suite_x509parse.data (hunk 11 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes - `dumpasn1` knows about the expected content for various ASN.1 types such as OIDs. The tests currently don't make an attempt to consistently provide valid content for those types because the parser doesn't inspect it.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes, let's have `05 00` here as in the other tests!

Comment (AndrzejKurek) [reply tests/suites/test_suite_x509parse.data]: This sounds a bit like providing not-fully-working data just because we know the parser implementation, which contradicts the assumption that we should provide all well-formed and valid fields apart from those that we're testing that should fail.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes, so far, the tests only attempt to always provide correct _structure_, i.e. expected sequence and nesting of ASN.1 structures outside the deliberate point of failure, but they don't attempt to go further than that and always provide valid _content_. Ideally, they should, but I stopped at this point because - you're right - I knew that this would be enough for the parsing flexibility I needed.

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: The valid IssuerID used here is an object with zero length, and that isn't really valid.

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: I would recommend adding a test where issuerUniqueID is present, but the version of certificate is v1.

Comment (AndrzejKurek) [reply tests/suites/test_suite_x509parse.data]: And the same for subjectUniqueID, and an extension with v2 or v1.

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: All these tests use invalid objects for issuerUniqueID and subjectUniqueID.

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: What length mismatch is present here?

Comment (AndrzejKurek) [inline tests/suites/test_suite_x509parse.data]: I would describe it in more detail - GeneralizedTime in notBefore field of Validity, and would suggest adding a second test for the same error but in notAfter field.

File: tests/suites/test_suite_x509parse.data (hunk 1 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: This is no longer relevant I think.

File: tests/suites/test_suite_x509parse.data (hunk 2 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes, so far the tests don't make an effort to use valid OIDs. I'd prefer to leave that for a separate PR.

File: tests/suites/test_suite_x509parse.data (hunk 3 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: The test is meant to test the presence of additional data after `notBefore` and `notAfter`; so, the data is correct, but the test description isn't very clear. I'll change it to "data remaining after 'notAfter'".

File: tests/suites/test_suite_x509parse.data (hunk 5 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: We are testing various ill-formed structures, too, but for the purpose of testing the behavior of the parser on invalid tags, it's ok to keep a well-formed structure.

File: tests/suites/test_suite_x509parse.data (hunk 6 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Yes, that's cleaner.

File: tests/suites/test_suite_x509parse.data (hunk 7 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Fixed.

File: tests/suites/test_suite_x509parse.data (hunk 8 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Fixed.

File: tests/suites/test_suite_x509parse.data (hunk 11 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Good idea!

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Done.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: They use empty IssuerID and SubjectID, which I thought is valid.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: It's deliberate that the tests use an empty OID here, but you're right in that it would be in the spirit of the PR to use valid ones. However, for the main purpose of the PR to allow modifying the order of our parsing steps, it's not necessary, because our name parsing routines accept any byte string as an OID.

I push that to the back for now and see if I find time. Would you be willing to approve the PR even without this change?

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Actually, the issuerID and subjectID can be any ASN.1 bitstring.

However, you make a good point here: Currently, our parsing module only checks the explicit outer tag, but entirely ignores the inner tag, which is supposed to be a BITSTRING tag.

That said, I'd prefer to only create an issue for now, because even though the standard is explicit that it should be a BITSTRING, that doesn't mean that there aren't CRTs in use which use other types, e.g. INT, and we shouldn't break users of such CRTs (e.g., even though the standard says the serial number must be an INTEGER, we except some other explicit tag there, too).

Are you OK with that?

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: @AndrzejKurek Sorry, forget what I said - `issuerID` and `subjectID` are _implicitly_ tagged, which means that the actual `BIT STRING` tag is omitted.

So, to summarize: Our parsing accepts any content for `issuerID` and `subjectID`, while it _should_ be a valid ASN.1 `BIT STRING` according to RFC 5280 (with a leading byte indicating the number of unused bits, followed by the actual contents).

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: There's data following the `extnValue` - I will clarify that in the test case description.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Fixed.

Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Done.

**Comment by hanno-arm** [comment]
@AndrzejKurek @Patater 

I believe to have addressed your issues, apart from (a) always using valid OIDs, (b) always using non-empty issuerID/subjectID. I would like to postpone these changes because (a) of time, and (b) it's not needed to fulfill the main purpose of this PR, which is allowing us to arbitrarily reorder our parsing routine; that's because we don't parse name OIDs, nor subjectID, nor issuerID.

Please let me know if you're ok with that.

**Comment by hanno-arm** [comment]
@Patater Do you think you'll have the time to re-review this?

**Comment by hanno-arm** [comment]
@AndrzejKurek @Patater Can you please re-review?

File: tests/suites/test_suite_x509parse.data [hunk 12]
```diff
@@ -1650,6 +1650,14 @@ X509 CRT ASN1 (TBS, inv v3Ext, SubjectAltName repeated outside Extensions)
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"3081dc3081c6a0030201028204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffffa100a200a321301f301d0603551d11041630148208666f6f2e7465737482086261722e74657374301d0603551d11041630148208666f6f2e7465737482086261722e74657374300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
 
+X509 CRT ASN1 (TBS, valid v3Ext in v1 CRT)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"3081b93081a3a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffffa321301f301d0603551d11041630148208666f6f2e7465737482086261722e74657374300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Why is this `MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`? I am having trouble seeing where the length mismatch is. Is it from an interpretation of the ASN.1 instead of the ASN.1 itself?

File: tests/suites/test_suite_x509parse.data [hunk 13]
```diff
@@ -1650,6 +1650,14 @@ X509 CRT ASN1 (TBS, inv v3Ext, SubjectAltName repeated outside Extensions)
...
+
+X509 CRT ASN1 (TBS, valid v3Ext in v2 CRT)
+depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
+x509parse_crt:"3081b93081a3a0030201018204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210ffffffffffffffffffffffffffffffff0202ffffa321301f301d0603551d11041630148208666f6f2e7465737482086261722e74657374300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH
```
Comment (Patater) [inline tests/suites/test_suite_x509parse.data]: Same here. I don't see why this would return length mismatch error.

File: tests/suites/test_suite_x509parse.data (hunk 12 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: @Patater When we think we should have reached the end of an ASN.1 structure but we haven't, we return `MBEDTLS_ERR_ASN1_LENGTH_MISMATCH`, which I think is the right choice of error code for this kind of situation. The case triggers an instance of that, because the v1 CRT parser doesn't look for extensions but expects to have parsed the entire TBS body, which isn't the case if anything follows (in particular a misplaced v3 extension).

File: tests/suites/test_suite_x509parse.data (hunk 13 above)
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: Same explanation.

**Comment by hanno-arm** [comment]
Thanks @Patater for reviewing, I believe I have addressed your comments. Could you give a final look?

File: tests/suites/test_suite_x509parse.data (hunk 12 above)
Comment (Patater) [reply tests/suites/test_suite_x509parse.data]: OK, thanks for explaining. This makes sense now. It's not that any of the lengths are wrong at the ASN.1 level, but at the CRT level. Thanks!

**Comment by hanno-arm** [comment]
@AndrzejKurek @Patater I rebased this on top of the current `development` and #2442. Could you please review?

**Comment by hanno-arm** [comment]
PR-head CI failure due to flaky DTLS test. Restarting.

**Comment by hanno-arm** [comment]
@AndrzejKurek @Patater 

Rebasing was quite laborious, which seems to be mainly because in the base PR #2442 there has been lots of case conversion (to downcase). To address this, I took the original version [x509_parse_bf_040619/b4f0db9d24b7430125da70f7102d8f50987ff7af](https://github.com/hanno-arm/mbedtls/tree/x509_parse_bf_040619) of this PR and created a zero-diff version [x509_parse_bf_with_lower_case_commits](https://github.com/hanno-arm/mbedtls/tree/x509_parse_bf_with_lower_case_commits) which pre-fixes the main commits of this PR by commits which only do the lower-case changes. Rebasing this branch on top of 3c03a881eb37b0ad63d7cdb6813e1106d5f30d74 / #2430 was still a bit of manual work, but easier: 
* First two conflicts arise because of additions of new tests in #2530 - just keep them.
* Next conflict comes from renaming of a test "X509 Certificate ASN1 (correct)" to "X509 Certificate ASN1 (well-formed)" - easy to fix.
* Next conflict is with "Shorten X.509 CRT parsing test names" because some tests for DER-encoded CRTs have been added in the meantime. Keep the these tests, but apply the name shortening.

The result of this should be similar to [x509_parse_bf_after_rebase](https://github.com/hanno-arm/mbedtls/tree/x509_parse_bf_after_rebase), which is very close to current state of this PR, modulo a few smaller changes - check the diff.

File: tests/suites/test_suite_x509parse.data [hunk 14]
```diff
@@ -1498,29 +1498,31 @@ X509 Certificate ASN1 (TBSCertificate, invalid SubjectPublicKeyinfo, check faile
 depends_on:MBEDTLS_RSA_C:MBEDTLS_SHA256_C
 x509parse_crt:"308196308180a0030201008204deadbeef300d06092a864886f70d01010b0500300c310a30080600130454657374301c170c303930313031303030303030170c303931323331323335393539300c310a30080600130454657374302a300d06092A864886F70D010101050003190030160210fffffffffffffffffffffffffffffffe0202ffff300d06092a864886f70d01010b0500030200ff":"":MBEDTLS_ERR_PK_INVALID_PUBKEY
 
-X509 Certificate ASN1 (TBSCertificate v3, Optional UIDs, Extensions not present)
+# We don't expect MBEDTLS_ERR_ASN1_UNEXPECTED_TAG here because the IssuerID is optional;
+# the stack is expected to ignore it and will hence end up with a length mismatch.
+X509 Certificate ASN1 (TBSCertificate, invalid IssuerID, invalid tag)
```
[the remark this answers is not in the document]
Comment (hanno-arm) [reply tests/suites/test_suite_x509parse.data]: You're right, the documentation is outdated. It would make sense in v2 Certificates which don't contain extensions, but it doesn't make sense here. I pushed a commit to fix this.

**Comment by hanno-arm** [comment]
@Patater I've addressed your comment. Please re-review.

@AndrzejKurek Could you also re-review, please?

**Comment by hanno-arm** [comment]
@Patater @AndrzejKurek Thank you very much for your review! I appreciate it a lot that you took the time to do this painful review.

**Comment by Patater** [comment]
@hanno-arm Thanks for the testing improvements.
